portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -17,30 +17,60 @@
 import com.google.gson.Gson;
 import java.util.ArrayList;
 import java.io.IOException;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private ArrayList<String> comments;
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-       
-    ArrayList<String> comments = new ArrayList<String>();
-    comments.add("First Comment");
-    comments.add("Second Comment");
-    comments.add("Third Comment");
 
-    // Convert string of hard-coded messages to JSON
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    comments = new ArrayList<String>();
+
+    for (Entity entity : results.asIterable()) {
+      String text = (String) entity.getProperty("text");
+      comments.add(text);
+    }
+
     String json = convertToJson(comments);
 
-    // Send JSON as response
     response.setContentType("application/json;");
     response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     
+    String text = request.getParameter("text-input");
+    String name = request.getParameter("name");
+
+    long timestamp = System.currentTimeMillis();
+
+    Entity taskEntity = new Entity("Comment");
+    taskEntity.setProperty("text", text);
```

Comment:
Since these string constants are used in multiple places, I recommend making them `private static final String`s instead to guard against typos and changes to the names.

Reply:
private static final here gives an error or illegal modifier for the parameter. I think I can only use final here?

Reply:
I think what Andrew means is declaring: 
```
private static final String textParameter = "text-input";
private static final String nameParameter = "name";
private static final String timestampParameter = "timestamp";
```
as static constants for the class. So adding these lines at line 33, outside of any function. That way you can change these lines to use that constant:
`String text = request.getParamater(textParameter);`
and later on:
`taskEntity.setProperty(textParameter, text);`

And likewise for the other parameters

Reply:
Ahh, I see. Makes sense!